File: php/src/OpenHealthhub/PhpFhirClient/CarePlanClient.php
<?php


namespace OpenHealthhub\PhpFhirClient;


use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRBackboneElement\FHIRAppointment\FHIRAppointmentParticipant;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRCanonical;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRContactPoint;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRContactPointSystem;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRExtension;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRHumanName;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRIdentifier;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRInstant;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRParticipationStatus;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRPeriod;
use DCarbone\PHPFHIRGenerated\R4\FHIRElement\FHIRReference;
use DCarbone\PHPFHIRGenerated\R4\FHIRResource\FHIRBundle;
use DCarbone\PHPFHIRGenerated\R4\FHIRResource\FHIRDomainResource\FHIRAppointment;
use DCarbone\PHPFHIRGenerated\R4\FHIRResource\FHIRDomainResource\FHIRCarePlan;
use DCarbone\PHPFHIRGenerated\R4\FHIRResource\FHIRDomainResource\FHIRPatient;

class CarePlanClient
{
    public function getCarePlan($id): FHIRCarePlan
    {
        $client = new FhirClient();
        $res = $client->get(sprintf('CarePlan/%s', $id));
        return new FHIRCarePlan($res);
    }

    public function searchCarePlans($planDefReference): FHIRBundle
    {
        $client = new FhirClient();
        $res = $client->search(sprintf('CarePlan?instantiates-canonical=PlanDefinition/%s', $planDefReference));
        return new FHIRBundle($res);
    }

    public function searchCarePlanWithPractitioners($id): FHIRBundle
    {
        $client = new FhirClient();
        $res = $client->search(sprintf('CarePlan?_id=%s&_include=CarePlan:care-team&_include=CareTeam:participant', $id));
        return new FHIRBundle($res);
    }

    public function createCarePlan(): FHIRCarePlan
    {
        $client = new FhirClient();
        $careplan = $this->createCarePlanResource();

        $res = $client->create('CarePlan', $careplan);
        return new FHIRCarePlan($res);
    }

    public function updateCarePlan(): FHIRCarePlan
    {
        $client = new FhirClient();
        $careplan = $this->createCarePlanResource();
        $careplan->setId("1");

        $res = $client->update('CarePlan', $careplan);
        return new FHIRCarePlan($res);
    }

    public function deleteCarePlan($id)
    {
        $client = new FhirClient();
        $res = $client->delete(sprintf('CarePlan/%s', $id));
    }

    private function createCarePlanResource(): FHIRCarePlan
    {
        $careplan = new FHIRCarePlan();

        $patient = new FHIRPatient();
        $patientName = new FHIRHumanName();
        $patientName->setText('Test Patient');
        $patientEmail = new FHIRContactPoint();
        $telecomType = new FHIRContactPointSystem();
        $telecomType->setValue('email');
        $patientEmail->setSystem($telecomType)->setValue('test@patient.ohh');
        $patientIdentifier = new FHIRIdentifier();
        $patientIdentifier->setSystem('urn:oid:2.16.840.1.113883.2.4.99')->setValue('1234');
        $patient->setId('patient')->addName($patientName)->addTelecom($patientEmail)->addIdentifier($patientIdentifier);
        $careplan->addContained($patient);
        $careplan->setSubject(new FHIRReference(['#patient']));

        $careplan->setInstantiatesCanonical([new FHIRCanonical('PlanDefinition/cca2eaf3-03a9-46c0-88c6-e0287917cea6')]);
        $period = new FHIRPeriod();
        $period->setStart([new FHIRInstant('2021-03-16T13:32:37.430+01:00')]);
        $careplan->setPeriod($period);
        return $careplan;
    }
}

File: java/src/main/java/com/openhealthhub/careplan/CarePlanClient.java
package com.openhealthhub.careplan;

import ca.uhn.fhir.model.api.Include;
import ca.uhn.fhir.rest.api.MethodOutcome;
import ca.uhn.fhir.rest.client.api.IGenericClient;
import com.openhealthhub.util.FhirUtil;
import org.hl7.fhir.instance.model.api.IAnyResource;
import org.hl7.fhir.r4.model.Bundle;
import org.hl7.fhir.r4.model.CanonicalType;
import org.hl7.fhir.r4.model.CarePlan;
import org.hl7.fhir.r4.model.CareTeam;
import org.hl7.fhir.r4.model.ContactPoint;
import org.hl7.fhir.r4.model.HumanName;
import org.hl7.fhir.r4.model.Identifier;
import org.hl7.fhir.r4.model.Patient;
import org.hl7.fhir.r4.model.Period;
import org.hl7.fhir.r4.model.Reference;

import java.time.Instant;
import java.util.Collections;
import java.util.Date;
import java.util.List;

public class CarePlanClient {

    private final IGenericClient client;

    public static void main(String... args) {
        String action = args.length == 0 ? "get" : args[0];
        CarePlanClient client = new CarePlanClient();

        switch (action) {
            case "create":
                FhirUtil.printResource(client.createCarePlan());
                return;
            case "search":
                FhirUtil.printResource(client.searchCarePlan());
                return;
            case "getInclude":
                FhirUtil.printResource(client.getCarePlanWithPractitioners());
                return;
            case "update":
                FhirUtil.printResource(client.updateCarePlan());
                return;
            case "delete":
                client.deleteCarePlan();
                return;
            default:
                FhirUtil.printResource(client.getCarePlan(args.length == 2 ? args[1] : "4"));
        }
    }

    public Bundle searchCarePlan() {
        return client.search()
                .forResource(CarePlan.class)
                .where(CarePlan.INSTANTIATES_CANONICAL.hasId("PlanDefinition/97f680b9-e397-4298-8c53-de62a284c806"))
                .and(CarePlan.PATIENT.hasChainedProperty(Patient.IDENTIFIER.exactly().identifier("1234")))
                .returnBundle(Bundle.class)
                .execute();
    }

    public Bundle getCarePlanWithPractitioners() {
        return client.search()
                .forResource(CarePlan.class)
                .where(IAnyResource.RES_ID.exactly().identifier("4"))
                .include(new Include("CarePlan:" + CarePlan.SP_CARE_TEAM))
                .include(new Include("CareTeam:" + CareTeam.SP_PARTICIPANT))
                .returnBundle(Bundle.class)
                .execute();
    }

    private void deleteCarePlan() {
        client.delete().resourceById("CarePlan", "1").execute();
    }

    private MethodOutcome updateCarePlan() {
        var carePlan = createCarePlanResource();
        carePlan.setId("1");

        return client.update().resource(carePlan).execute();
    }

    CarePlanClient() {
        client = FhirUtil.createClient();
    }

    CarePlan getCarePlan(String id) {
        return client.read()
                .resource(CarePlan.class)
                .withId(id)
                .execute();
    }

    MethodOutcome createCarePlan() {
        CarePlan carePlan = createCarePlanResource();

        return client.create().resource(carePlan).execute();

    }

    private CarePlan createCarePlanResource() {
        CarePlan carePlan = new CarePlan();
        carePlan.setPeriod(new Period().setStart(Date.from(Instant.now())));

        Patient patient = new Patient();
        patient.setId("patient");
        HumanName humanName = new HumanName();
        humanName.setText("Test Patient");
        patient.setName(Collections.singletonList(humanName));

        ContactPoint email = new ContactPoint();
        email.setSystem(ContactPoint.ContactPointSystem.EMAIL);
        email.setValue("test@patient.ohh");
        patient.addTelecom(email);

        Identifier programPatientId = new Identifier();
        programPatientId.setSystem("urn:oid:2.16.840.1.113883.2.4.99");
        programPatientId.setValue("1234");
        patient.addIdentifier(programPatientId);

        carePlan.addContained(patient);
        carePlan.setSubject(new Reference("#patient"));
        carePlan.setInstantiatesCanonical(List.of(new CanonicalType("PlanDefinition/cca2eaf3-03a9-46c0-88c6-e0287917cea6")));
        return carePlan;
    }
}
